### PR TITLE
[export] allow user input mutation in aot_export

### LINKED ISSUE #119330: aot_export should support user input mutation
```
RuntimeError: 
Found following user inputs located at [0, 2, 3, 5] are mutated. This is currently banned in the aot_export workflow.
If you need this functionality, please file a github issue.
```
Guess I'm filing the GitHub issue 😛. Export has to [work around](https://fburl.com/kv0xgxfz) this atm, would be nice to just support it in `aot_export` proper.

### PR DESCRIPTION
**Pull request by suo**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #119356

This PR enables input mutation in aot_export by removing the guard and ensuring that the GraphSignature is properly wired up.

This allows to undo the gross hack in torch.export where we lift user inputs to buffers in order to get around aot_export upstream support. It also makes input mutation work properly for non-strict mode.

Mutations on inputs that require_grad are still banned (I added a test for a non-parameter input as well, just to make sure).

Differential Revision: [D53507440](https://our.internmc.facebook.com/intern/diff/D53507440/)

### CODE REVIEW

File: torch/export/_trace.py [hunk 1]
```diff
@@ -439,6 +356,11 @@ def _export_non_strict(
             trace_joint=False,
             pre_dispatch=pre_dispatch,
         )
+    # TODO unfortunately preserving graph-level metadata is not
```
Comment (bdhirsh) [inline torch/export/_trace.py:359]: 🤔 in theory the stuff Sherlock added for preserving nn stack traces from dynamo also means that node meta should be preserved

Comment (bdhirsh) [reply torch/export/_trace.py:359]: Oh, graph level meta

**Review by bdhirsh — APPROVED** [review]
Lgtm!

**Review by zhxchen17 — CHANGES_REQUESTED** [review]
change looks good, but I prefer the previous node name convention which was regressed...

File: test/export/test_export.py [hunk 2]
```diff
@@ -1207,14 +1207,14 @@ def forward(self, x):
             str(gm.code).strip(),
             """\
 def forward(self, arg_0):
-    l_x_, = fx_pytree.tree_flatten_spec(([arg_0], {}), self._in_spec)
+    arg7_1, = fx_pytree.tree_flatten_spec(([arg_0], {}), self._in_spec)
```
Comment (zhxchen17) [inline test/export/test_export.py:1210]: I think for debugging purposes I'd like to keep the original node names like `l_x_`

File: torch/export/_trace.py [hunk 3]
```diff
@@ -329,88 +328,6 @@ def _export_to_torch_ir(
...
-
-    for node in list(gm_torch_level.graph.nodes):
-        if node.op == "placeholder":
-            assert len(node.users) == 0
-            gm_torch_level.graph.erase_node(node)
-    gm_torch_level.recompile()
-    return user_input_names
-
-
-def _lift_buffers_to_user_inputs(
```
Comment (zhxchen17) [inline torch/export/_trace.py:356]: but removing this is kinda nice.

**Review by zhxchen17 — APPROVED** [review]
after offline discussion, let's improve node names as a follow up after this diff

**Comment by thiagocrepaldi** [comment]
@titaiwangms FYI

**Review by titaiwangms — APPROVED** [review]

File: torch/onnx/_internal/exporter.py [hunk 4]
```diff
@@ -799,9 +799,9 @@ def model_signature(self) -> Optional[torch.export.ExportGraphSignature]:
                               target='my_buffer2', persistent=True),
                     InputSpec(kind=<InputKind.BUFFER: 3>, arg=TensorArgument(name='arg5_1'),
                               target='my_buffer1', persistent=True),
-                    InputSpec(kind=<InputKind.USER_INPUT: 1>, arg=TensorArgument(name='l_x_'),
+                    InputSpec(kind=<InputKind.USER_INPUT: 1>, arg=TensorArgument(name='arg6_1'),
```
Comment (thiagocrepaldi) [inline torch/onnx/_internal/exporter.py:802]: Isnt it possible to preserve the name instead?

File: test/onnx/test_fx_op_consistency.py [hunk 5]
```diff
@@ -1629,32 +1617,32 @@ def skip_torchlib_forward_compatibility(
...
         reason="does not match node name",
     ),
-    xfail(
+    skip(
         "signal.windows.hann",
         model_type=pytorch_test_common.TorchModelType.TORCH_EXPORT_EXPORTEDPROGRAM,
         reason="does not match node name",
     ),
-    xfail(
+    skip(
```
Comment (thiagocrepaldi) [inline test/onnx/test_fx_op_consistency.py:1645]: why changing xfail by skip?

File: torch/onnx/_internal/exporter.py (hunk 4 above)
Comment (suo) [reply torch/onnx/_internal/exporter.py:802]: trickier with the current approach. It also was not guaranteed that these names would look the way they do if strict=False. I plan to followup with a more comprehensive input processing pass that should name these more nicely.

File: test/onnx/test_fx_op_consistency.py (hunk 5 above)
Comment (suo) [reply test/onnx/test_fx_op_consistency.py:1645]: could use some help investigating this; for some reason if I left `xfail` then I was getting "unexpected success", but if I removed the xfail I was getting a failure…

Comment (thiagocrepaldi) [reply test/onnx/test_fx_op_consistency.py:1645]: @liqunfu @titaiwangms @BowenBao was this the behavior you noticed this week too? what was the fix?

Comment (thiagocrepaldi) [reply test/onnx/test_fx_op_consistency.py:1645]: > could use some help investigating this; for some reason if I left `xfail` then I was getting "unexpected success", but if I removed the xfail I was getting a failure…

@suo we will get back on this for you

File: torch/onnx/_internal/exporter.py (hunk 4 above)
Comment (thiagocrepaldi) [reply torch/onnx/_internal/exporter.py:802]: > trickier with the current approach. It also was not guaranteed that these names would look the way they do if strict=False. I plan to followup with a more comprehensive input processing pass that should name these more nicely.

is there an issue filed for this? would love to know more about it

Comment (suo) [reply torch/onnx/_internal/exporter.py:802]: No, was just the conversation on this PR. Feel free to file an issue which we will keep updated

File: test/onnx/test_fx_op_consistency.py (hunk 5 above)
Comment (titaiwangms) [reply test/onnx/test_fx_op_consistency.py:1645]: Put it down to the list SKIP_XFAIL_SUBTESTS with xfail should work. Ithink I met this kind of weird op tests before.

File: torch/onnx/_internal/exporter.py (hunk 4 above)
Comment (thiagocrepaldi) [reply torch/onnx/_internal/exporter.py:802]: https://github.com/pytorch/pytorch/issues/119484 thanks

**Pull request closed without merging**